src/lib/io/sol-uart-impl-linux.c

```diff
@@ -32,49 +33,70 @@
 #include "sol-uart.h"
 #include "sol-util-internal.h"
 #include "sol-util-file.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
 #define FD_ERROR_FLAGS (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)
 
 struct sol_uart {
+    struct sol_fd *fd_handler;
+    const void *user_data;
+    void (*rx_cb)(void *data, struct sol_uart *uart);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    struct sol_ptr_vector pending_blobs;
+    size_t written; //How many bytes We've written so far for the first element of pending_blobs
     int fd;
-    struct {
-        void *rx_fd_handler;
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void *tx_fd_handler;
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length, tx_index;
-    } async;
 };
 
+
 static bool
-uart_rx_callback(void *data, int fd, uint32_t active_flags)
+uart_fd_handler_callback(void *data, int fd, uint32_t active_flags)
 {
     struct sol_uart *uart = data;
-
-    if (!uart->async.rx_cb)
-        return true;
+    struct sol_blob *blob;
+    ssize_t r;
 
     if (active_flags & FD_ERROR_FLAGS) {
         SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
         return true;
     }
+
     if (active_flags & SOL_FD_FLAGS_IN) {
-        uint8_t c; /* Backing storage for next sol_buffer */
-        struct sol_buffer buf = SOL_BUFFER_INIT_FLAGS(&c, sizeof(c),
-            SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
-        int status = sol_util_fill_buffer(uart->fd, &buf, 1);
-        if (status > 0)
-            uart->async.rx_cb((void *)uart->async.rx_user_data, uart, c);
+        uart->rx_cb((void *)uart->user_data, uart);
+        return true;
+    }
+
+    //SOL_FD_FLAGS_OUT
+
+    blob = sol_ptr_vector_get_nocheck(&uart->pending_blobs, 0);
+    r = write(uart->fd, (char *)blob->mem + uart->written,
+        blob->size - uart->written);
+
+    if (r < 0) {
+        if (errno != EAGAIN) {
+            int aux_errno = errno;
+            SOL_WRN("Could not write at the UART fd: %d - Reason:%s", uart->fd,
+                sol_util_strerrora(aux_errno));
+            if (uart->tx_cb)
+                uart->tx_cb((void *)uart->user_data, uart, blob, -aux_errno);
+        }
+        return false;
+    }
+    uart->written += r;
+    if (uart->written == blob->size) {
+        if (uart->tx_cb)
+            uart->tx_cb((void *)uart->user_data, uart, blob, 0);
```

Comment:
if `sol_uart_close()` is called from `tx_cb`, then it crashes. You should postpone this as the last thing to do (of course move `sol_blob_unref(blob)` as well).